src/parseYarnDescriptor.js

```diff
@@ -0,0 +1,41 @@
+/**
+ * A regular expression for parsing a yarn berry descriptor.
+ */
+const PARSE_REGEX = /(^@?[^/]+?\/?[^@/]+?)@(?:.*:)*(.+)/;
+
+/**
+ * @typedef {Object} ParsedDescriptor
+ * @property {string} packageName
+ * @property {string} version
+ */
+
+/**
+ * Parses a yarn berry descriptor into an object.
+ *
+ * For example `@material/ripple@npm:1.0.0` will be parsed to:
+ * ```
+ * {
+ *  package: '@material/ripple',
+ *  version: '1.0.0',
+ * }
+ * ```
+ *
+ * Supported formats:
+ * - `package@version`
+ * - `@scope/package@version`
+ * - `package@protocol:version`
+ * - `@scope/package@protocol:version`
+ *
+ * @param {string} descriptor - The yarn berry descriptor to parse.
+ * @returns {ParsedDescriptor} The parsed descriptor.
+ */
+function parseYarnDescriptor(descriptor) {
```

Comment:
```suggestion
function parseBerryDescriptor(descriptor) {
```
Maybe? I was worried this regex would slow down non-berry users at first glance. Speaking of that, how's the perf with a large berry workspace like yours?

Reply:
Performance looks good. Only adds ~4% increase in time in a large project

Before PR:
```
  Time (mean ± σ):      4.028 s ±  0.101 s    [User: 3.634 s, System: 0.342 s]
  Range (min … max):    3.935 s …  4.207 s    10 runs
```
After PR:
```
  Time (mean ± σ):      4.121 s ±  0.152 s    [User: 3.748 s, System: 0.328 s]
  Range (min … max):    3.975 s …  4.453 s    10 runs
```

Reply:
As for the suggestion, this function parses both yarn v1 and berry descriptors. At the time of calling it we don't know which version of yarn lockfile we are parsing. Unless of course we interpret the "metadata" value in the parse result, but that's a bit messy.

I can clarify that this parses both descriptors in the JSDoc

Reply:
Sweet, makes sense!